extensions/pkg/controller/dnsrecord/reconciler.go

```diff
@@ -0,0 +1,217 @@
+	if err := controllerutils.EnsureFinalizer(ctx, r.reader, r.client, dns, FinalizerName); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err := r.statusUpdater.Processing(ctx, dns, operationType, "Reconciling the dnsrecord"); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	r.logger.Info("Starting the reconciliation of dnsrecord", "dnsrecord", dns.Name)
+	if err := r.actuator.Reconcile(ctx, dns, cluster); err != nil {
+		_ = r.statusUpdater.Error(ctx, dns, extensionscontroller.ReconcileErrCauseOrErr(err), operationType, "Error reconciling dnsrecord")
+		return extensionscontroller.ReconcileErr(err)
+	}
+
+	if err := r.statusUpdater.Success(ctx, dns, operationType, "Successfully reconciled dnsrecord"); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *reconciler) restore(ctx context.Context, dns *extensionsv1alpha1.DNSRecord, cluster *extensionscontroller.Cluster) (reconcile.Result, error) {
+	if err := controllerutils.EnsureFinalizer(ctx, r.reader, r.client, dns, FinalizerName); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err := r.statusUpdater.Processing(ctx, dns, gardencorev1beta1.LastOperationTypeRestore, "Restoring the dnsrecord"); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	r.logger.Info("Starting the restoration of dnsrecord", "dnsrecord", dns.Name)
+	if err := r.actuator.Restore(ctx, dns, cluster); err != nil {
+		_ = r.statusUpdater.Error(ctx, dns, extensionscontroller.ReconcileErrCauseOrErr(err), gardencorev1beta1.LastOperationTypeRestore, "Error restoring dnsrecord")
+		return extensionscontroller.ReconcileErr(err)
+	}
+
+	if err := r.statusUpdater.Success(ctx, dns, gardencorev1beta1.LastOperationTypeRestore, "Successfully restored dnsrecord"); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err := extensionscontroller.RemoveAnnotation(ctx, r.client, dns, v1beta1constants.GardenerOperation); err != nil {
+		return reconcile.Result{}, fmt.Errorf("error removing annotation from DNSRecord: %+v", err)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *reconciler) migrate(ctx context.Context, dns *extensionsv1alpha1.DNSRecord, cluster *extensionscontroller.Cluster) (reconcile.Result, error) {
+	if err := r.statusUpdater.Processing(ctx, dns, gardencorev1beta1.LastOperationTypeMigrate, "Migrating the dnsrecord"); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	r.logger.Info("Starting the migration of dnsrecord", "dnsrecord", dns.Name)
+	if err := r.actuator.Migrate(ctx, dns, cluster); err != nil {
+		_ = r.statusUpdater.Error(ctx, dns, extensionscontroller.ReconcileErrCauseOrErr(err), gardencorev1beta1.LastOperationTypeMigrate, "Error migrating dnsrecord")
+		return extensionscontroller.ReconcileErr(err)
+	}
+
+	if err := r.statusUpdater.Success(ctx, dns, gardencorev1beta1.LastOperationTypeMigrate, "Successfully migrated dnsrecord"); err != nil {
```

Comment:
One reason to keep this behavior would be that once the finalizers are removed, if the object is already deleted it immediately disappears and the status is not updateable any longer, the attempt to update it will just result in an error.

> I guess, this can be a source a leaked resources (-> seed namespaces) during control plane migration, right?

If the finalizer is not removed successfully, then deleting the shoot namespace will hang for some time. As far as I can tell there are no other consequences so this should not be critical. For comparison, in the bad case scenario we will leave entire shoot control planes hanging around unused until they are cleaned up properly (manually at first, automatically later). To close the gap, we will have to implement the forcefull "purge" of such control planes at some point anyway.

In short, I think this is rather minor, although of course it could be fixed.

Reply:
Thanks for the detailed thoughts.

> To close the gap, we will have to implement the forcefull "purge" of such control planes at some point anyway.

Makes sense, got it :)

Reply:
The current flow might cause a bigger problem.
During the migration phase, the gardenlet will annotate resources with the migrate annotation and wait for them to be migrated successfully. After that it tries to delete the resources. So if it happened that the finalzers do not get removed due to an error, but the status does and shows a successful migrate operation, we actually run the risk of deleting resources.

Reply:
@plkokanov Yes, you are right, I somehow keep forgetting that we have active extension controllers to act upon these resources. I assume changing the order, e.g. first removing the finalizers, and only then updating the status, will fix this issue? This needs to be done consistently for all extensions, so I would suggest to do it in a different PR.